retrystorage/retry_storage.go

```diff
@@ -0,0 +1,194 @@
+package retrystorage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+)
+
+const (
+	// defaultMaxAttempts is the default number of attempts of each storate
+	// operations before it finally fails.
+	defaultMaxAttempts = 3
+)
+
+type stoppingBackOff struct {
+	attempts    int
+	MaxAttempts int
+	Underlying  backoff.BackOff
+}
+
+func (s *stoppingBackOff) NextBackOff() time.Duration {
+	if s.attempts >= s.MaxAttempts {
+		return backoff.Stop
+	}
+	s.attempts++
+	return s.Underlying.NextBackOff()
+}
+
+func (s *stoppingBackOff) Reset() {
+	s.attempts = 0
+	s.Underlying.Reset()
+}
+
+type RetryStorageConfig struct {
+	// Dependencies.
+
+	Logger     micrologger.Logger
+	Underlying microstorage.Storage
+
+	// Settings.
+
+	NewBackOffFunc func() backoff.BackOff
+}
+
+func DefaultConfig() RetryStorageConfig {
+	return RetryStorageConfig{
+		// Dependencies.
+
+		Logger:     nil, // Required.
+		Underlying: nil, // Required.
+
+		// Settings.
+
+		NewBackOffFunc: func() backoff.BackOff {
+			return &stoppingBackOff{
+				MaxAttempts: defaultMaxAttempts,
+				Underlying:  backoff.NewExponentialBackOff(),
+			}
+		},
+	}
+}
+
+type RetryStorage struct {
+	logger         micrologger.Logger
+	underlying     microstorage.Storage
+	newBackOffFunc func() backoff.BackOff
+}
+
+func New(config RetryStorageConfig) (*RetryStorage, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger is empty")
+	}
+	if config.Underlying == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Underlying is empty")
+	}
+	if config.NewBackOffFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.NewBackOffFunc is empty")
+	}
+
+	s := &RetryStorage{
+		logger:         config.Logger,
+		underlying:     config.Underlying,
+		newBackOffFunc: config.NewBackOffFunc,
+	}
+
+	return s, nil
+}
+
+func (s *RetryStorage) Create(ctx context.Context, key, value string) error {
+	b := s.newBackOffFunc()
+	op := func() error {
+		err := s.underlying.Create(ctx, key, value)
+		if microstorage.IsInvalidKey(err) || microstorage.IsNotFound(err) {
+			return backoff.Permanent(err)
+		}
+		return err
+	}
+	notify := func(err error, delay time.Duration) {
+		s.logger.Log("warn", fmt.Sprintf("retrying create key=%s delay=%s", key, delay), "err", fmt.Sprintf("%#v", err))
```

Comment:
IMO we should use the nouns here, not the verbs. There is no verb for `error`. It would be `failing` or something but that does not work. Same for `info`. We are used to the noun `error` already. The nouns further down the road would be `debug`, `info` and `warning`.